base/essentials.jl

```diff
@@ -243,10 +216,14 @@ function typename(a::Union)
 end
 typename(union::UnionAll) = typename(union.body)
 
-convert(::Type{T}, x::T) where {T<:Tuple{Any, Vararg{Any}}} = x
-convert(::Type{Tuple{}}, x::Tuple{Any, Vararg{Any}}) = throw(MethodError(convert, (Tuple{}, x)))
-convert(::Type{T}, x::Tuple{Any, Vararg{Any}}) where {T<:Tuple} =
-    (convert(tuple_type_head(T), x[1]), convert(tuple_type_tail(T), tail(x))...)
```

Comment:
Do you have test cases where these fail?

Reply:
The actual problem that I'm trying to workaround with this PR's `fieldtype`-based implementation is e.g.:

```julia
julia> using Cassette

julia> Cassette.@context Ctx
ERROR: invalid redefinition of constant Ctx
Stacktrace:
 [1] top-level scope at /Users/jarrettrevels/.julia/dev/Cassette/src/macros.jl:19

julia> Cassette.@overdub(Ctx(), convert(Tuple{Vararg{Float64}}, (1,2)))
ERROR: type DataType has no field body
Stacktrace:
 [1] execution at /Users/jarrettrevels/.julia/dev/Cassette/src/overdub.jl:9 [inlined]
 [2] macro expansion at /Users/jarrettrevels/.julia/dev/Cassette/src/overdub.jl:171 [inlined]
 [3] overdub_recurse at /Users/jarrettrevels/.julia/dev/Cassette/src/overdub.jl:201 [inlined]
 [4] overdub_execute at /Users/jarrettrevels/.julia/dev/Cassette/src/overdub.jl:35 [inlined]
 [5] getproperty at ./sysimg.jl:15 [inlined]
 [6] overdub_recurse(::Cassette.Context{getfield(Main, Symbol("##CtxName#350")),Nothing,Cassette.NoPass,Nothing,Nothing}, ::typeof(getproperty), ::Type{Vararg{Float64,N} where N}, ::Symbol) at /Users/jarrettrevels/.julia/dev/Cassette/src/overdub.jl:0
 [7] overdub_execute at /Users/jarrettrevels/.julia/dev/Cassette/src/overdub.jl:35 [inlined]
 [8] unwrap_unionall at ./essentials.jl:179 [inlined]
 [9] overdub_recurse(::Cassette.Context{getfield(Main, Symbol("##CtxName#350")),Nothing,Cassette.NoPass,Nothing,Nothing}, ::typeof(Base.unwrap_unionall), ::Type{Vararg{Float64,N} where N}) at /Users/jarrettrevels/.julia/dev/Cassette/src/overdub.jl:0
 [10] overdub_execute at /Users/jarrettrevels/.julia/dev/Cassette/src/overdub.jl:35 [inlined]
 [11] unwrapva at ./essentials.jl:224 [inlined]
 [12] overdub_recurse(::Cassette.Context{getfield(Main, Symbol("##CtxName#350")),Nothing,Cassette.NoPass,Nothing,Nothing}, ::typeof(Base.unwrapva), ::Type{Vararg{Float64,N} where N}) at /Users/jarrettrevels/.julia/dev/Cassette/src/overdub.jl:0
 [13] overdub_execute at /Users/jarrettrevels/.julia/dev/Cassette/src/overdub.jl:35 [inlined]
 [14] tuple_type_head at ./essentials.jl:154 [inlined]
 [15] overdub_recurse(::Cassette.Context{getfield(Main, Symbol("##CtxName#350")),Nothing,Cassette.NoPass,Nothing,Nothing}, ::typeof(Base.tuple_type_head), ::Type{Tuple{Vararg{Float64,N} where N}}) at /Users/jarrettrevels/.julia/dev/Cassette/src/overdub.jl:0
 [16] overdub_execute at /Users/jarrettrevels/.julia/dev/Cassette/src/overdub.jl:35 [inlined]
 [17] convert at ./essentials.jl:248 [inlined]
 [18] overdub_recurse(::Cassette.Context{getfield(Main, Symbol("##CtxName#350")),Nothing,Cassette.NoPass,Nothing,Nothing}, ::typeof(convert), ::Type{Tuple{Vararg{Float64,N} where N}}, ::Tuple{Int64,Int64}) at /Users/jarrettrevels/.julia/dev/Cassette/src/overdub.jl:0
 [19] overdub_execute at /Users/jarrettrevels/.julia/dev/Cassette/src/overdub.jl:35 [inlined]
 [20] #19 at /Users/jarrettrevels/.julia/dev/Cassette/src/macros.jl:94 [inlined]
 [21] overdub_recurse(::Cassette.Context{getfield(Main, Symbol("##CtxName#350")),Nothing,Cassette.NoPass,Nothing,Nothing}, ::getfield(Main, Symbol("##19#20"))) at /Users/jarrettrevels/.julia/dev/Cassette/src/overdub.jl:0
 [22] top-level scope at none:0
```

which seems to be from `tuple_type_head`/`tuple_type_tail` hitting some kindtype subtyping/inference issues that I don't fully understand. With this PR, I get the correct answer e.g.

```
julia> Cassette.@overdub(Ctx(), convert(Tuple{Vararg{Float64}}, (1,2)))
(1.0, 2.0)
```